src/Microsoft.AspNet.Hosting/Internal/HostingApplication.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Hosting.Server;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Hosting.Internal
+{
+    public class HostingApplication : IHttpApplication
+    {
+        private readonly IServiceProvider _applicationServices;
+        private readonly RequestDelegate _application;
+        private readonly ILogger _logger;
+        private readonly DiagnosticSource _diagnosticSource;
+        private readonly IHttpContextFactory _httpContextFactory;
+        private IDisposable _logScope;
+
+        public HostingApplication(
+            IServiceProvider applicationServices,
+            RequestDelegate application,
+            ILogger logger,
+            DiagnosticSource diagnosticSource,
+            IHttpContextFactory httpContextFactory)
+        {
+            _applicationServices = applicationServices;
+            _application = application;
+            _logger = logger;
+            _diagnosticSource = diagnosticSource;
+            _httpContextFactory = httpContextFactory;
+        }
+
+        public object CreateHttpContext(object contextFeatures)
+        {
+            var httpContext = _httpContextFactory.Create((IFeatureCollection)contextFeatures);
+            
+            httpContext.ApplicationServices = _applicationServices;
+
+            if (_diagnosticSource.IsEnabled("Microsoft.AspNet.Hosting.BeginRequest"))
+            {
+                _diagnosticSource.Write("Microsoft.AspNet.Hosting.BeginRequest", new { httpContext = httpContext });
+            }
+
+            _logScope = _logger.RequestScope(httpContext);
```

Comment:
Change the interface definition to 

```cs
public interface IHttpApplication<THttpContext>
{
  THttpContext CreateHttpContext(IFeatureCollection requestFeatures);
  Task InvokeAsync(THttpContext httpContext);
  void DisposeHttpContext(THttpContext httpContext);
}
//and
public interface IServer : IDisposable
{
  IFeatureCollection Features { get; }
  void Start<THttpContext>(IHttpApplication<THttpContext> app);
}
```
Then in the case of this Hosting assembly: create a struct `Q` which has two public fields, `IDisposable LoggerScope` and `HttpContext HttpContext`. So it is a zero-allocation value-type (tuple).

Then `HttpApplication : IHttpApplication<Q>` and you call `server.Start<Q>(app);`. The `Start<Q>` will probably also tell you that the Q can be inferred.